### PR TITLE
List View: allow expanding and collapsing of nested blocks

### PR DESCRIPTION
**Pull request by gwwar**
Part of https://github.com/WordPress/gutenberg/issues/26141 to add the ability to expand and collapse parts of list view.

Updates in this PR include click support for expand/collapse and left and right keyboard events, as noted in https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html. A left keypress on a expanded item should collapse it, and a second keypress should navigate to its parent. Similarly a right keypress on a collapsed item should expand it.

I plan on splitting up work for List View expand/collapse in two parts:

- **This PR** will focus on basic UI + click/keyboard events.
- **A follow up PR** will add animations for slide in and out for the tree grid rows. Work here is somewhat non-trivial due to the nature of the markup (nested items aren't children but siblings) and no native support for transitions on [auto dimensions](https://css-tricks.com/using-css-transitions-auto-dimensions/). We'll need JS to animate this. Ideally react spring should be updated before we add another animation here. See https://github.com/WordPress/gutenberg/pull/30979

Due to related interest, I also wrote up a quick issue around if Tree Grid still makes sense for this component https://github.com/WordPress/gutenberg/issues/32294

https://user-images.githubusercontent.com/1270189/119903244-4a733680-befd-11eb-9704-d819408f648f.mp4

### Requested Feedback 👀 

~~The main thing that sticks out is **state management** and **keyboard/focus handling**. Right now keyboard handling is implemented at the Tree Grid level, which makes triggering callbacks at the branch level very difficult. I couldn't think of a good way of preserving nested collapsed state without resorting to a heavier data solution like a data store, or something a bit hacky like the click() call I made. Let me know if folks had thoughts around that.~~ I think I have this mostly sorted with useReducer, but I can also make a data store here if folks would prefer that instead.

### Testing Instructions

- Navigate to the Post or Site Editor, and click to display the list view.
- Verify that clicking on the chevron toggles display of nested items
- Hidden items should not be rendered in the DOM
- Toggle state of child items should be preserved.
- Other list views like in the Navigation block should still work/make sense

https://user-images.githubusercontent.com/1270189/119904700-97f0a300-beff-11eb-8f78-3e4fa7b837cd.mp4

### CODE REVIEW

**Comment by jasmussen** [comment]
Took this for a quick spin, even if it's still a work in progress. This is what I see:

list view

Already now, this is an amazing improvement, it's incredibly better. I have a few design metrics that I'll help push so the various buttons are just right. This one also shows an ellipsis menu:

Screenshot 2021-05-26 at 09 22 14

Even just in this state, it surfaces some interesting questions that I'm happy to chat through. For example: if you select a deeply nested block in the canvas, which is inside a collapsed group, does it uncollapse the group? And also, should the groups that you collapse stay collapsed even if you close and reopen the list view? Should every top level item start collapsed? All excellent questions that I imagine we can tune across a few PRs and not something we have to get right in this one. All to say, amazing work.

**Comment by gwwar** [comment]
Thanks for adding images @jasmussen, I ran out of time last week ✨ 

> Even just in this state, it surfaces some interesting questions that I'm happy to chat through. 

💯 I wanted to chat through this too!

One thing that was much more jarring is that the primary action **selects the block**. This is not much of an issue with a mouse, but with a keyboard this means that **focus is lost on the side panel**, and a second <kbd>enter</kbd> creates a new paragraph, rather than toggling primary action.

> For example: if you select a deeply nested block in the canvas, which is inside a collapsed group, does it uncollapse the group? And also, should the groups that you collapse stay collapsed even if you close and reopen the list view? 

There's a demo in treegrid on behavior https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html, where the first click selects, and the second click expands. I'm not sure I like the feel of the tristate toggle myself, but we could do something similar.

https://user-images.githubusercontent.com/1270189/119690466-925f6400-bdfe-11eb-82a6-4e158641f97e.mp4

Another thing that's different with the current implementation is that we're using a single td cell, and we need to be able to select nested items. This means that we need to reserve the primary action <kbd>enter</kbd> for selecting the block rather than expand/collapsing the item. Reserving Left and right keyboard events to expand and toggle feels okay.

> And also, should the groups that you collapse stay collapsed even if you close and reopen the list view?

I would expect it to retain remain collapsed, but also provide an expand all button/open all folders to the selected node. It's actually cheaper to not keep track of non-visible toggle state, so it's pretty simple to implement if we feel otherwise. We could also let other folks weigh in and modify this in follow up PRs.

>  Should every top level item start collapsed?

It'd be much more performant for pages with lots of nested items to do this, but the tradeoff is more overhead for what is this/what do I do? I do think we'd want to hold off on trying this until we have implemented an open all nested items button, or open all items to our selected node target button.

> All excellent questions that I imagine we can tune across a few PRs and not something we have to get right in this one. All to say, amazing work.

To not let this get too out of scope, I'll aim to add the necessary missing keyboard events and try merging with behavior as is, unless folks spot anything that we should have for the first pass.

**Comment by gwwar** [comment]
Looking at some of the keyboard handling, it probably makes more sense to try and put the expand/collapse state + handling in the treegrid component, I'll give that a try next.

**Comment by jasmussen** [comment]
> One thing that was much more jarring is that the primary action selects the block. This is not much of an issue with a mouse, but with a keyboard this means that focus is lost on the side panel, and a second enter creates a new paragraph, rather than toggling primary action.

Just to be sure, this is describing existing trunk behavior also, correct? It isn't specific to the addition of the expand/collapse chevron? Also CC: @Copons @jameskoster for some of the conversations around the list view. There may be prior art in the inserter, where if you tab to select a block and press Enter, it inserts it but doesn't move focus there. Whereas tab and press ⌘ + Enter, it inserts it and moves focus there. If I recall the conversation correctly, that is a known pattern for screen readers and interaction shortcuts.

> Reserving Left and right keyboard events to expand and toggle feels okay.

👍 👍

> We could also let other folks weigh in and modify this in follow up PRs.

Definitely good to keep this PR as small as possible, I only got excited about all the potential followups! And yes, I do think it'd be nice that a group you intentionally collapsed stays collapsed across your entire session, despite opening and closing the list view.

**Comment by jameskoster** [comment]
> One thing that was much more jarring is that the primary action selects the block. This is not much of an issue with a mouse, but with a keyboard this means that focus is lost on the side panel, and a second enter creates a new paragraph, rather than toggling primary action.

I'm pretty sure this is the issue @copons has surfaced a number of times: there's no streamlined way for one to move focus back to list view after selecting a block. The 'workaround' is that you need to close + reopen it which is one step more than it should be. One solution might be separate keyboard shortcuts for opening/closing rather than a single toggle, but that's probably something to explore separately.

**Comment by jasmussen** [comment]
By the way, let me know when you're ready for me to push some polish here to the paddings (just little things) ✨

**Comment by gwwar** [comment]
> By the way, let me know when you're ready for me to push some polish here to the paddings (just little things) ✨

@jasmussen I'll give you a ping when I'm ready, I'm still sorting out the best place to handle state 👍

**Comment by gwwar** [comment]
I added a quick work around for the keyboard support. @jasmussen feel free to add commits to the branch. 👍

File: packages/block-editor/src/components/block-navigation/tree.js
Comment (gwwar) [inline packages/block-editor/src/components/block-navigation/tree.js]: The main thing that sticks out is state management and keyboard/focus handling. Right now keyboard handling is implemented at the Tree Grid level, which makes triggering callbacks at the branch level very difficult. I couldn't think of a good way of preserving nested collapsed state without resorting to a heavier data solution like a data store, or something a bit hacky like the click() call I made. Let me know if folks had thoughts around that.

File: packages/block-editor/src/components/block-navigation/expander.js
Comment (gwwar) [inline packages/block-editor/src/components/block-navigation/expander.js:5]: Ideally in the future we don't need a component like this, and can instead use CSS rules to point at an SVG image.

**Comment by jasmussen** [comment]
I pushed a little polish:

Screenshot 2021-05-28 at 09 52 05

This is based on the metrics in https://github.com/WordPress/gutenberg/issues/32294#issuecomment-850199386, and involved creating a new "chevron-small" icon. I believe these utility icons are useful, just like we have a close and close small icon, for specific contexts like these.

File: packages/block-editor/src/components/block-navigation/style.scss
Comment (jasmussen) [inline packages/block-editor/src/components/block-navigation/style.scss:124]: CC: @jameskoster, the icon was scaled down to 20px, so it didn't show true pixels. I shifted the metrics around a bit, but would appreciate you taking a look when you have a chance, thank you.

Comment (jameskoster) [reply packages/block-editor/src/components/block-navigation/style.scss:124]: Screenshot 2021-05-28 at 10 41 48

Looks good to me.

**Comment by jameskoster** [comment]
One small issue I spotted: If you select and collapse a parent, it would be nice if the bottom border radii matched the top. Here's how it looks currently:

Screenshot 2021-05-28 at 10 42 45

---

@jasmussen I appreciate the smaller chevron, but am wondering if it could be even smaller? I'm still feeling some tension between the block icon and the chevron. Here's a slightly downsized version (keeping the same overall footprint):

Screenshot 2021-05-28 at 10 55 56

And an even more downsized version (1px stroke):

Screenshot 2021-05-28 at 10 55 26

What do you think?

**Comment by jasmussen** [comment]
I'd be on board with a smaller one, but I'd keep it 1.5px so as to not introduce new visual ingredients, which it'd feel like at 1px.

**Comment by jameskoster** [comment]
It's a shame, the 1px feels better to me in practise, but I understand the reluctance to add more ingredients :D 

I pushed a smaller version of the 1.5px chevron to try. It's a little larger (filesize) because I outlined the stroke.

**Comment by jameskoster** [comment]
Figma updated: https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=9490%3A85

**Comment by jasmussen** [comment]
Nice one, thanks Jay!

File: packages/block-editor/src/components/block-navigation/tree.js
Comment (gwwar) [reply packages/block-editor/src/components/block-navigation/tree.js]: Oh hmm, 🤔 maybe I can see if using React `useReducer` helps here at all.

File: packages/block-editor/src/components/block-navigation/tree.js
Comment (gwwar) [inline packages/block-editor/src/components/block-navigation/tree.js:26]: We only record state when someone toggles an expand/collapse chevron. 

One case I haven't covered is when blocks are deleted, we might have some extra state left over eg:

```jsx
state = { 
   'block-client-id-a': false,
   'block-client-id-b': true,
   'deleted-client-id': false
};
```

It is possible to prune this by dispatching an event as a useEffect side effect when `clientIdsTree`, or something similar like listening to a block delete event. I chose to leave this out since it does feel a bit unnecessary since this is likely to remain pretty small. Let me know if others feel otherwise and I can add handling.

**Comment by jameskoster** [comment]
@jasmussen how do you feel about aligning the chevron centrally with the icon above it? We lose a little horizontal space, but it feels a bit neater to my eye:

| Aligned | Not aligned (current) |
| --- | --- |
| Screenshot 2021-06-02 at 14 28 56 | Screenshot 2021-06-02 at 14 09 07 |

**Comment by jasmussen** [comment]
I don't love it. Actually it feels more noisy to me because it makes it feel more separate rather than part of the item itself.

**Comment by jameskoster** [comment]
Hah, that's why I like it. I don't feel any anxiety when expanding/collapsing because the button _feels_ easier to target. Especially when all rows are collapsed to the same level like so: 

| Aligned | Not aligned |
| --- | --- |
| Screenshot 2021-06-02 at 14 32 12 | Screenshot 2021-06-02 at 14 32 50 |

I can train my cursor to the larger block icon, then just move down the rows vertically and expand/collapse without any horizontal movement required.

FWIW this seems to be a fairly common pattern, as seen in MacOS:

Screenshot 2021-06-02 at 14 40 47

Material:

Screenshot 2021-06-02 at 14 49 25

And of course our best friend Figma :D 

Screenshot 2021-06-02 at 14 41 37

It's a fairly small css change so I'll push, and we can easily revert if it doesn't feel good.

**Comment by jasmussen** [comment]
Here's what I see:

new

I don't hate it!

I still don't love it, but I don't hate it :D and because it's a small tweak, I'm happy to try it.

**Comment by shaunandrews** [comment]
I've left this branch running for most of the day on my site and I absolutely love the folding in the List View. It makes working with blocks much easier, especially if you're building or working with patterns.

I didn't find any bugs or unexpected behavior. 🚢

File: packages/block-editor/src/components/block-navigation/style.scss
Comment (gwwar) [inline packages/block-editor/src/components/block-navigation/style.scss]: Is there a formula for the pixel indentation?

Comment (jasmussen) [reply packages/block-editor/src/components/block-navigation/style.scss]: Probably n * $icon-size, let me know if that works.

File: packages/block-editor/src/components/block-navigation/block-select-button.js
Comment (Copons) [inline packages/block-editor/src/components/block-navigation/block-select-button.js]: Wonder if we should separate `onClick` with a new prop like `onToggleExpanded`.

`onClick` implies listening for the click on the entire component's surface, which... is the case but we prevent triggering the toggle in `toggleExpandOrSelectBlock` with the `forceToggle` parameter.

It looks a bit convoluted and unclear to me, and I think that two separate props, one for selecting the block and one for toggling the branch, would make the component easier to consume. 🤔

File: packages/components/src/tree-grid/index.js
Comment (Copons) [inline packages/components/src/tree-grid/index.js:130]: I love the `left` behaviour!! It works so effortlessly and exactly as I'd expect! ✨ 

I've got a suggestion for the `right` behaviour.
Nothing happens if you hit `right` again after expanding a branch.
This felt a bit awkward to me, and I've realized that's because I'm used to how the VSCode file tree works.

Hitting `right` again after expanding would move the selection to the first child.
If the first child is a branch, `right` would either expand it, or move to its first child.
Once the first child is a leaf, hitting `right` wouldn't do anything anymore.
(Hope I've been clear 😅 )

This is imho super handy to traverse down the tree, in a specular way to the upward-traversal of `left`.

**Review by Copons — COMMENTED** [review]
This works great @gwwar! Thanks for working on it! ✨

**Comment by gwwar** [comment]
As a heads up, I'm going to rebase this branch with trunk

File: packages/components/src/tree-grid/index.js
Comment (gwwar) [reply packages/components/src/tree-grid/index.js:130]: Right arrow doing nothing in the rightmost column is from treegrid behavior. I'd certainly be open to adding more functionality if folks agree with the behavior.

https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html

Screen Shot 2021-06-11 at 8 42 18 AM

Comment (Copons) [reply packages/components/src/tree-grid/index.js:130]: I see, thanks for the explanation!
I'd certainly agree with the behaviour I've described earlier. 😄

File: packages/block-editor/src/components/block-navigation/style.scss
Comment (gwwar) [reply packages/block-editor/src/components/block-navigation/style.scss]: Thanks! @jasmussen @jameskoster I updated this in https://github.com/WordPress/gutenberg/pull/32117/commits/7fab9c8e484c081112f9cd98a357653b2e69c4c6 Let me know if that's too confusing, and we can keep the expanded form otherwise.

Screen Shot 2021-06-11 at 9 50 14 AM

File: packages/block-editor/src/components/block-navigation/block-select-button.js
Comment (gwwar) [reply packages/block-editor/src/components/block-navigation/block-select-button.js]: Thanks, I'll see if I can make it clearer. I think I might have left it that way from experimenting with a few different state approaches.

Comment (gwwar) [reply packages/block-editor/src/components/block-navigation/block-select-button.js]: I tried teasing those two functions out in https://github.com/WordPress/gutenberg/pull/32117/commits/1b8b13d37b1f1306d9291e0a1cd6f8dfbb4ee321. There's a bit of prop passing, but let me know what you think.

**Comment by chrisvanpatten** [comment]
This looks really slick! Hope to see it ship!

File: packages/components/src/tree-grid/index.js
Comment (talldan) [inline packages/components/src/tree-grid/index.js]: This and the similar line below couples the generic treegrid implementation to only work with elements that have a data-block attribute (blocks). I think it should ideally work for any kind of tree regardless of the content.

I'm not too precious about it (the whole thing could end up being deleted given the other discussions about moving away from tree grid in #32294 .)

But if there's a quick fix it would be good. Maybe pass the element back to the List View code or something, and that code can instead be responsible for accessing the `dataset.block` on the element.

File: packages/block-editor/src/components/block-navigation/expander.js
Comment (talldan) [inline packages/block-editor/src/components/block-navigation/expander.js:16]: The implementation as a pseudo-element in the w3 example was really surprising to me. 

I guess that makes it completely hidden from the accessibility tree, in which case the span could have `aria-hidden="true"` specified to replicate that.

It might be worth adding some comments to the component explaining why this is implemented like it is—I imagine many devs with good intentions might be tempted to change it to a `Button` which wouldn't be good given it's already rendered in a `Button`.

File: packages/block-editor/src/components/block-navigation/tree.js
Comment (talldan) [reply packages/block-editor/src/components/block-navigation/tree.js:26]: Yeah, agreed, it's probably not a big issue. I doubt this will ever become a huge amount of data.

I have ideas but I think it's an effort vs. reward question more than anything:
- BlockNavigation could receive stable block client ids from a selector (e.g. from `getBlockOrder`), and reconcile it against the `expanded` state in an effect.
- It could be possible to move the state to the treegrid component entirely. Could each `TreeGridRow` manage its own state? I think it'd then have to render its own expander component, but that seems potentially ok. I might be missing some complications.

File: packages/block-editor/src/components/block-navigation/style.scss
Comment (jasmussen) [reply packages/block-editor/src/components/block-navigation/style.scss]: Looks good to me! Thanks.

File: packages/components/src/tree-grid/index.js
Comment (talldan) [reply packages/components/src/tree-grid/index.js:130]: Looks like that's how a tree view works:
https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-22

So it might be determined by how https://github.com/WordPress/gutenberg/issues/32294 is progressed. I'm not sure how that can be moved forwards, but probably worth coming to a decision soon.

**Review by talldan — COMMENTED** [review]
Looks and works great. Comments are all pretty minor.

**Comment by gwwar** [comment]
> It could be possible to move the state to the treegrid component entirely. Could each TreeGridRow manage its own state? I think it'd then have to render its own expander component, but that seems potentially ok. I might be missing some complications.

Right we could try to simplify state by making this a boolean value on a child component versus a map in List View. Two things that stopped me from going with that approach were:

Ideally, we do not render children in the DOM if it's collapsed. The List View mirrors the full document state in some use cases, which is a performance sensitive area. 

https://user-images.githubusercontent.com/1270189/121925168-7c540d80-ccf1-11eb-9ca6-b6b7942660f5.mp4

Persistence of toggle stage (Eg if we collapse child items, and then toggle a parent, do we retain state of those toggled child elements, or does it default to all toggled open?).

https://user-images.githubusercontent.com/1270189/121925250-92fa6480-ccf1-11eb-91d9-66aef0c47232.mp4

File: packages/block-editor/src/components/block-navigation/expander.js
Comment (gwwar) [reply packages/block-editor/src/components/block-navigation/expander.js:16]: Good call, I've updated this in https://github.com/WordPress/gutenberg/pull/32117/commits/f4b5a5f3527628d8969938ab3024189a0ef37348

For folks 👀  we can see the pseudo element in https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html
 
Screen Shot 2021-06-14 at 9 25 57 AM

File: packages/components/src/tree-grid/index.js
Comment (gwwar) [reply packages/components/src/tree-grid/index.js]: I think that's pretty fair. I've updated this in 90dd73f92b01dfba21a12b05d32400e0d8b145a4 to pass back the row element instead

File: packages/components/src/tree-grid/index.js
Comment (gwwar) [reply packages/components/src/tree-grid/index.js:130]: Thanks, I revisited this and I caught some behavior I misimplemented. I should be keeping focus in the rightmost cell when expanded instead of looping to the first focus item. This is updated in https://github.com/WordPress/gutenberg/pull/32117/commits/7065ab6e560363e551b9852d4de4f4117835c8cc 

We can see this in the navigation list view:

Before:


https://user-images.githubusercontent.com/1270189/121932756-16b84f00-ccfa-11eb-9ecc-a3a01d3816fa.mp4


After:

https://user-images.githubusercontent.com/1270189/121932775-2041b700-ccfa-11eb-8ac7-1cb8f8c55219.mp4

**Comment by gwwar** [comment]
I think this should be good to go after another round of 👀 cc @Copons @talldan. Thanks for everyone's feedback so far!

**Review by talldan — APPROVED** [review]
Fantastic addition, great work @gwwar.

For a follow-up, I did notice that when closing and reopening list view, the expanded/collapsed state isn't retained, I guess a side-effect of the component no longer being rendered and the component state being lost.

**Comment by gwwar** [comment]
Thanks for taking another look @talldan!

> For a follow-up, I did notice that when closing and reopening list view, the expanded/collapsed state isn't retained, I guess a side-effect of the component no longer being rendered and the component state being lost.

To support that we can move component state into a data store. Shouldn't be too bad to do, but we'll need to make sure that other list view instances are not sharing state.

**Comment by shaunandrews** [comment]
🥳

**Comment by talldan** [comment]
> To support that we can move component state into a data store. Shouldn't be too bad to do, but we'll need to make sure that other list view instances are not sharing state.

Another thing I wondered is whether this state should be part of the block editor store. Looking through that store, most of the actions/selectors relate directly to an actual block or block list, while this is a view on a block list, so it does feel slightly different to the main purpose of that store. 🤔

**Comment by mtias** [comment]
Awesome to see this, great work everyone! Really happy with all the iterations and refinements on this elements, it's becoming more and more useful and powerful.

**Pull request merged**
